services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java

```diff
@@ -172,14 +175,40 @@ public TokenEndpoint(TokenManager tokenManager, RealmModel realm, EventBuilder e
     @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
     @POST
     public Response processGrantRequest() {
+        // grant request needs to be run in a retriable transaction as concurrent execution of this action can lead to
+        // exceptions on DBs with SERIALIZABLE isolation level.
+        Object result = KeycloakModelUtils.runJobInRetriableTransaction(this.session.getKeycloakSessionFactory(), kcSession -> {
+            try {
+                RealmModel realmModel = kcSession.realms().getRealm(realm.getId());
+                kcSession.getContext().setRealm(realmModel);
+                // create another instance of the endpoint that will be run within the new session.
+                Resteasy.pushContext(KeycloakSession.class, kcSession);
+                TokenEndpoint other = new TokenEndpoint(new TokenManager(), realmModel, new EventBuilder(realmModel, kcSession, clientConnection));
+                ResteasyProviderFactory.getInstance().injectProperties(other);
+                return other.processGrantRequestInternal();
```

Comment:
I like the idea of creating a new instance. I wonder if the context should be restored afterwards? See below for a (untested) suggestion.

```suggestion
                KeycloakSession oldSession = Resteasy.getContextData(KeycloakSession.class);
                try {
                    Resteasy.pushContext(KeycloakSession.class, kcSession);
                    TokenEndpoint other = new TokenEndpoint(new TokenManager(), realmModel, new EventBuilder(realmModel, kcSession, clientConnection));
                    ResteasyProviderFactory.getInstance().injectProperties(other);
                    return other.processGrantRequestInternal();
                } finally {
                    Resteasy.pushContext(KeycloakSession.class, oldSession);
                }

```

Reply:
Generally this is a good suggestion. Perhaps even better would be to introduce a `Resteasy.popContext` operation.

In this particular case, it is not needed though since the response is not a `Stream`, and the correct session is closed in the request filter. 

Therefore I'm approving this PR as is. Restoring the context can be done as follow-up.